File: 2021/day02/puzzle.py
from dataclasses import dataclass

input = open("input.txt", "r").read().splitlines()


@dataclass
class Pos:
    horizontal: int
    depth: int

    def add(self, add):
        self.horizontal += add.horizontal
        self.depth += add.depth

    def __str__(self):
        return str(self.horizontal * self.depth)


def part_one():
    position = Pos(0, 0)
    for command in input:
        direction, length = command.split(" ")
        if direction == "forward":
            position.add(Pos(int(length), 0))
        elif direction == "up":
            position.add(Pos(0, -int(length)))
        else:
            position.add(Pos(0, int(length)))

    print(position)


@dataclass
class PosWithAim(Pos):
    aim: int

    def add(self, add):
        self.horizontal += add.horizontal
        self.depth += add.depth
        self.aim += add.aim


def part_two():
    position = PosWithAim(0, 0, 0)
    for command in input:
        direction, arg = command.split(" ")
        length = int(arg)
        if direction == "forward":
            position.add(PosWithAim(length, position.aim * length, 0))
        elif direction == "up":
            position.add(PosWithAim(0, 0, -length))
        else:
            position.add(PosWithAim(0, 0, length))

    print(position)


part_one()
part_two()

File: 2021/day02/puzzle.mjs
import { promises } from "fs";

const input = (await promises.readFile("./input.txt", { encoding: "utf8" }))
  .split("\n")
  .slice(0, -1)
  .map((v) => v.split(" "));

const mul = (x, y) => x * y
const add = ([x, y], [x2, y2]) => [x + x2, y + y2]

const doCommand = ([direction, length]) => {
  switch (direction) {
    case "forward":
      return [parseInt(length), 0]
    case "down":
      return [0, parseInt(length)]
    case "up":
      return [0, -parseInt(length)]
  }
}

const navigate = (commands) => commands.reduce((position, command) => add(position, doCommand(command)), [0, 0])

console.log(`Part 1 answer: ${mul(...navigate(input))}`)
